package org.curtinfrc.frc2026.subsystems.hoodedshooter;

import choreo.util.ChoreoAllianceFlipUtil;
import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.geometry.Pose3d;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.geometry.Rotation3d;
import edu.wpi.first.math.geometry.Transform3d;
import edu.wpi.first.math.geometry.Translation2d;
import edu.wpi.first.math.interpolation.InterpolatingDoubleTreeMap;
import edu.wpi.first.math.kinematics.ChassisSpeeds;
import edu.wpi.first.wpilibj.Alert;
import edu.wpi.first.wpilibj.Alert.AlertType;
import edu.wpi.first.wpilibj2.command.Command;
import edu.wpi.first.wpilibj2.command.SubsystemBase;
import edu.wpi.first.wpilibj2.command.button.Trigger;
import java.util.function.Supplier;
import org.curtinfrc.frc2026.Constants;
import org.curtinfrc.frc2026.Constants.Mode;
import org.curtinfrc.frc2026.drive.Drive;
import org.curtinfrc.frc2026.sim.BallSim;
import org.curtinfrc.frc2026.util.FieldConstants;
import org.curtinfrc.frc2026.util.LoggedTunableNumber;
import org.littletonrobotics.junction.Logger;

public class HoodedShooter extends SubsystemBase {
  public static final double MOTOR_WARNING_TEMP = 60;
  public static final double WHEEL_DIAMETER =
      (Constants.robotType == Constants.RobotType.COMP) ? 0.1 : 0.101;
  public static final int SHOOTER_MOTOR_NUMBER =
      (Constants.robotType == Constants.RobotType.COMP) ? 4 : 4;
  public static final int HOOD_MOTOR_NUMBER =
      (Constants.robotType == Constants.RobotType.COMP) ? 2 : 1;
  public static final Translation2d HUB_LOCATION =
      ChoreoAllianceFlipUtil.flip(FieldConstants.Hub.topCenterPoint.toTranslation2d());

  public static final double READY_SHOOTER_VELOCITY_TOLERANCE = 1.0;
  public static final double READY_HOOD_POSITION_TOLERANCE = 1.0;
  public static final double READY_ROBOT_ROTATION_TOLERANCE = 5.0;

  public static final InterpolatingDoubleTreeMap DISTANCE_TO_SHOOTER_VELOCITY =
      new InterpolatingDoubleTreeMap();
  public static final InterpolatingDoubleTreeMap DISTANCE_TO_HOOD_ANGLE =
      new InterpolatingDoubleTreeMap();
  public static final InterpolatingDoubleTreeMap DISTANCE_TO_BALL_FLIGHT_TIME =
      new InterpolatingDoubleTreeMap();

  private final HoodIO hoodIO;
  private final HoodIOInputsAutoLogged hoodInputs = new HoodIOInputsAutoLogged();

  private final ShooterIO shooterIO;
  private final ShooterIOInputsAutoLogged shooterInputs = new ShooterIOInputsAutoLogged();

  private final Supplier<Pose2d> robotPose;
  private final Supplier<ChassisSpeeds> robotVelocity;

  private final LoggedTunableNumber tunableHoodSetpoint =
      new LoggedTunableNumber("HoodSetpoint", 90);
  private final LoggedTunableNumber tunableShooterSetpoint =
      new LoggedTunableNumber("ShooterSetpoint", 26);
  private double shooterTarget = 0;
  private double hoodTarget = 0;

  private final Alert[] hoodMotorDisconnectedAlerts = new Alert[HOOD_MOTOR_NUMBER];
  private final Alert[] hoodMotorTempAlerts = new Alert[HOOD_MOTOR_NUMBER];
  private final Alert[] shooterMotorDisconnectedAlerts = new Alert[SHOOTER_MOTOR_NUMBER];
  private final Alert[] shooterMotorTempAlerts = new Alert[SHOOTER_MOTOR_NUMBER];

  public final Trigger hoodedShooterReady =
      new Trigger(
              () -> {
                double hoodPosition = hoodInputs.positionRotations * 360;
                boolean hoodReady =
                    Math.abs(hoodTarget - hoodPosition) <= READY_HOOD_POSITION_TOLERANCE;
                boolean shooterReady =
                    Math.abs(shooterTarget - shooterInputs.velocityMetresPerSecond)
                        <= READY_SHOOTER_VELOCITY_TOLERANCE;
                return hoodReady && shooterReady;
              })
          .debounce(0.1);

  public HoodedShooter(
      HoodIO hoodIO,
      ShooterIO shooterIO,
      Supplier<Pose2d> robotPose,
      Supplier<ChassisSpeeds> robotVelocity) {
    this.shooterIO = shooterIO;
    this.hoodIO = hoodIO;
    this.robotPose = robotPose;
    this.robotVelocity = robotVelocity;

    DISTANCE_TO_SHOOTER_VELOCITY.put(2.45, 14.1035);
    DISTANCE_TO_HOOD_ANGLE.put(2.45, 81.0);
    DISTANCE_TO_BALL_FLIGHT_TIME.put(2.45, 1.117);

    DISTANCE_TO_SHOOTER_VELOCITY.put(3.2, 14.1035);
    DISTANCE_TO_HOOD_ANGLE.put(3.2, 74.0);
    DISTANCE_TO_BALL_FLIGHT_TIME.put(3.2, 1.05);

    DISTANCE_TO_SHOOTER_VELOCITY.put(4.15, 16.7);
    DISTANCE_TO_HOOD_ANGLE.put(4.15, 70.0);
    DISTANCE_TO_BALL_FLIGHT_TIME.put(4.15, 1.267);

    // DISTANCE_TO_SHOOTER_VELOCITY.put(5.11, 19.5);
    // DISTANCE_TO_HOOD_ANGLE.put(5.11, 65.0);
    // DISTANCE_TO_BALL_FLIGHT_TIME.put(5.11, 1.2);

    for (int motor = 0; motor < HOOD_MOTOR_NUMBER; motor++) {
      hoodMotorDisconnectedAlerts[motor] =
          new Alert("Hood motor " + String.valueOf(motor) + " disconnected.", AlertType.kError);
      hoodMotorTempAlerts[motor] =
          new Alert(
              "Hood motor " + String.valueOf(motor) + " temperature above 60°C.",
              AlertType.kWarning);
    }
    for (int motor = 0; motor < SHOOTER_MOTOR_NUMBER; motor++) {
      shooterMotorDisconnectedAlerts[motor] =
          new Alert("Shooter motor " + String.valueOf(motor) + " disconnected.", AlertType.kError);
      shooterMotorTempAlerts[motor] =
          new Alert(
              "Shooter motor " + String.valueOf(motor) + " temperature above 60°C.",
              AlertType.kWarning);
    }
  }

  @Override
  public void periodic() {
    hoodIO.updateInputs(hoodInputs);
    shooterIO.updateInputs(shooterInputs);
    Logger.processInputs("Hood", hoodInputs);
    Logger.processInputs("Shooter", shooterInputs);
    Logger.recordOutput("HoodedShooter/hoodedShooterReady", hoodedShooterReady.getAsBoolean());
    Logger.recordOutput("HoodedShooter/hoodTarget", hoodTarget);
    Logger.recordOutput("HoodedShooter/shooterTarget", shooterTarget);
    Logger.recordOutput("HoodedShooter/hoodPositionDegrees", hoodInputs.positionRotations * 360);
    Logger.recordOutput(
        "HoodedShooter/distanceFromHub",
        HUB_LOCATION.minus(robotPose.get().getTranslation()).getNorm());

    for (int motor = 0; motor < hoodMotorDisconnectedAlerts.length; motor++) {
      hoodMotorDisconnectedAlerts[motor].set(!hoodInputs.motorsConnected[motor]);
      hoodMotorTempAlerts[motor].set(hoodInputs.motorTemperatures[motor] > MOTOR_WARNING_TEMP);
    }
    for (int motor = 0; motor < shooterMotorDisconnectedAlerts.length; motor++) {
      shooterMotorDisconnectedAlerts[motor].set(!shooterInputs.motorsConnected[motor]);
      shooterMotorTempAlerts[motor].set(
          shooterInputs.motorTemperatures[motor] > MOTOR_WARNING_TEMP);
    }
  }

  public Translation2d getVirtualTargetLocation(Supplier<Translation2d> location) {
    double realDistanceLength = location.get().minus(robotPose.get().getTranslation()).getNorm();
    Translation2d robotVel =
        new Translation2d(
            robotVelocity.get().vxMetersPerSecond, robotVelocity.get().vyMetersPerSecond);
    double airTime = DISTANCE_TO_BALL_FLIGHT_TIME.get(realDistanceLength);

    Translation2d hubCompensationOffset = robotVel.times(airTime);
    Translation2d compensatedHubLocation = location.get().plus(hubCompensationOffset);
    return (realDistanceLength > 1) ? compensatedHubLocation : location.get();
  }

  public Command shootAtTarget(Supplier<Translation2d> shotLocation) {
    return run(
        () -> {
          Translation2d compensatedHubLocation = getVirtualTargetLocation(shotLocation);

          double compensatedDistanceLength =
              compensatedHubLocation.minus(robotPose.get().getTranslation()).getNorm();

          if (Constants.tuningMode) {
            hoodTarget = tunableHoodSetpoint.get();
            shooterTarget = tunableShooterSetpoint.get();
          } else {
            hoodTarget = DISTANCE_TO_HOOD_ANGLE.get(compensatedDistanceLength);
            shooterTarget = DISTANCE_TO_SHOOTER_VELOCITY.get(compensatedDistanceLength);
          }

          double target =
              Drive.angleToLocation(this.getVirtualTargetLocation(shotLocation), robotPose.get());
          double robotAngle =
              robotPose.get().getRotation().rotateBy(Rotation2d.k180deg).getRadians();

          // shoot from other side
          // if (Math.abs(robotAngle) > Rotation2d.kCCW_90deg.getRadians()) {
          //   hoodTarget =
          //       Math.max(180 - hoodTarget, HoodIODev.FORWARD_LIMIT_ROTATIONS * 360); // clamp
          // value
          // }

          if (Math.abs(target - robotAngle) < READY_ROBOT_ROTATION_TOLERANCE) {
            hoodIO.setPosition(hoodTarget / 360);
            shooterIO.setVelocity(shooterTarget, hoodedShooterReady.getAsBoolean());
          } else {
            hoodIO.setVoltage(0);
            shooterIO.setVoltage(0);
          }

          if (Constants.getMode() == Mode.SIM) {
            shooterIO.addSimBall(
                new BallSim(
                    shooterTarget,
                    Rotation2d.fromDegrees(hoodTarget + 90),
                    new Pose3d(robotPose.get())
                        .plus(new Transform3d(0.2, 0.0, 0.3, Rotation3d.kZero))));
          }
        });
  }

  public Command setHoodPosition(double position) {
    return run(
        () -> {
          hoodIO.setPosition(position);
        });
  }

  public Command setHoodVoltage(double voltage) {
    return run(() -> hoodIO.setVoltage(voltage));
  }

  public Command stopHood() {
    return run(() -> hoodIO.setVoltage(0));
  }

  public Command setShooterVoltage(double voltage) {
    return run(() -> shooterIO.setVoltage(voltage));
  }

  public Command stopShooter() {
    return run(() -> shooterIO.setVoltage(0));
  }

  public Command setShooterVelocity(double velocityMetresPerSecond) {
    return run(
        () -> shooterIO.setVelocity(velocityMetresPerSecond, hoodedShooterReady.getAsBoolean()));
  }

  public Command setHoodedShooterPositionAndVelocity(
      double position, double velocityMetresPerSecond) {
    return run(
        () -> {
          hoodIO.setPosition(position);
          shooterIO.setVelocity(velocityMetresPerSecond, hoodedShooterReady.getAsBoolean());
        });
  }

  public Command stopHoodedShooter() {
    return run(
        () -> {
          hoodIO.setVoltage(0);
          shooterIO.setVelocity(0, false);
        });
  }
}
